kernel/names.ml

```diff
@@ -951,11 +951,71 @@ module GlobRef = struct
     | VarRef v1, VarRef v2 -> Id.equal v1 v2
     | (ConstRef _ | IndRef _ | ConstructRef _ | VarRef _), _ -> false
 
+  let global_eq_gen eq_cst eq_ind eq_cons x y =
+    x == y ||
+    match x, y with
+    | ConstRef cx, ConstRef cy -> eq_cst cx cy
+    | IndRef indx, IndRef indy -> eq_ind indx indy
+    | ConstructRef consx, ConstructRef consy -> eq_cons consx consy
+    | VarRef v1, VarRef v2 -> Id.equal v1 v2
+    | (VarRef _ | ConstRef _ | IndRef _ | ConstructRef _), _ -> false
+
+  let global_ord_gen ord_cst ord_ind ord_cons x y =
+    if x == y then 0
+    else match x, y with
+    | VarRef v1, VarRef v2 -> Id.compare v1 v2
+    | VarRef _, _ -> -1
+    | _, VarRef _ -> 1
+    | ConstRef cx, ConstRef cy -> ord_cst cx cy
+    | ConstRef _, _ -> -1
+    | _, ConstRef _ -> 1
+    | IndRef indx, IndRef indy -> ord_ind indx indy
+    | IndRef _, _ -> -1
+    | _ , IndRef _ -> 1
+    | ConstructRef consx, ConstructRef consy -> ord_cons consx consy
+
+  let global_hash_gen hash_cst hash_ind hash_cons gr =
+    let open Hashset.Combine in
+    match gr with
+    | ConstRef c -> combinesmall 1 (hash_cst c)
+    | IndRef i -> combinesmall 2 (hash_ind i)
+    | ConstructRef c -> combinesmall 3 (hash_cons c)
+    | VarRef id -> combinesmall 4 (Id.hash id)
+
 end
 
 type global_reference = GlobRef.t
 [@@ocaml.deprecated "Alias for [GlobRef.t]"]
 
+(* By default, [global_reference] are ordered on their canonical part *)
+
+module RefOrdered = struct
+  open Constant.CanOrd
+  type t = GlobRef.t
+  let compare gr1 gr2 =
+    GlobRef.global_ord_gen compare ind_ord constructor_ord gr1 gr2
+  let equal gr1 gr2 = GlobRef.global_eq_gen equal eq_ind eq_constructor gr1 gr2
+  let hash gr = GlobRef.global_hash_gen hash ind_hash constructor_hash gr
+end
+
+module RefOrdered_env = struct
+  open Constant.UserOrd
+  type t = GlobRef.t
+  let compare gr1 gr2 =
+    GlobRef.global_ord_gen compare ind_user_ord constructor_user_ord gr1 gr2
+  let equal gr1 gr2 =
+    GlobRef.global_eq_gen equal eq_user_ind eq_user_constructor gr1 gr2
+  let hash gr = GlobRef.global_hash_gen hash ind_user_hash constructor_user_hash gr
+end
+
+module Refmap = HMap.Make(RefOrdered)
+module Refset = Refmap.Set
```

Comment:
I think current trend is to use `GlobRef.Map` etc...

Reply:
Note that for constants and inductives, we still use the old scheme. The reason is probably that there are two different notions of maps, one based on canonical names and the other on user names.